reference/reference.go

```diff
@@ -0,0 +1,280 @@
+	dgst, s := p.parseDigest(s)
+	if len(s) > 0 {
+		p.err = ErrReferenceInvalidFormat
+	}
+	if p.err != nil {
+		return nil, p.err
+	}
+
+	if tag != "" && dgst != "" {
+		return NewTagDigestReference(repository, tag, digest.Digest(dgst)), nil
+	}
+	if dgst != "" {
+		return NewDigestReference(repository, digest.Digest(dgst)), nil
+	}
+	if tag != "" {
+		return NewTagReference(repository, tag), nil
+	}
+	return NewRepositoryReference(repository), nil
+}
+
+// Parse parses s and returns a syntactically valid Reference.
+// If an error was encountered it is returned, along with a nil Reference.
+func Parse(s string) (Reference, error) {
+	return new(parser).parse(s)
+}
+
+// DigestReference represents a reference of the form `repository@sha256:deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`.
+// Implements the Reference interface.
+type DigestReference struct {
+	repository string
+	digest     digest.Digest
+}
+
+// Repository returns the repository part.
+func (r DigestReference) Repository() string { return r.repository }
+
+// String returns the full string reference.
+func (r DigestReference) String() string {
+	return r.repository + "@" + string(r.digest)
+}
+
+// NewDigestReference returns an initialized DigestReference.
+func NewDigestReference(repo string, digest digest.Digest) DigestReference {
+	return DigestReference{
+		repo,
+		digest,
+	}
+}
+
+// TagReference represents a reference of the form `repository:tag`.
+// Implements the Reference interface.
+type TagReference struct {
+	repository string
+	tag        string
+}
+
+// Repository returns the repository part.
+func (r TagReference) Repository() string { return r.repository }
+
+// String returns the full string reference.
+func (r TagReference) String() string {
+	return r.repository + ":" + r.tag
+}
+
+// NewTagReference returns an initialized TagReference.
+func NewTagReference(repo string, tag string) TagReference {
+	return TagReference{
+		repo,
+		tag,
+	}
+}
+
+// TagDigestReference represents a reference of the form `repository:tag@sha256:deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`
+// with both tags and digests. Implements the Reference interface.
+type TagDigestReference struct {
+	repository string
+	tag        string
+	digest     digest.Digest
+}
+
+// Repository returns the repository part.
+func (r TagDigestReference) Repository() string { return r.repository }
+
+// String returns the full string reference.
+func (r TagDigestReference) String() string {
+	return r.repository + ":" + r.tag + "@" + string(r.digest)
+}
+
+// NewTagDigestReference returns an initialized TagDigestReference.
+func NewTagDigestReference(repository, tag string, digest digest.Digest) TagDigestReference {
+	return TagDigestReference{
+		repository,
+		tag,
+		digest,
+	}
+}
+
+// RepositoryReference represents a reference of the form `repository`, without tags nor digests.
+// Implements the Reference interface.
+type RepositoryReference struct {
```

Comment:
Wondering if we should just a protected string type for this, a la `digest.Digest`. It can still implement `reference.Reference`.